### PR TITLE
fix: fuel persistency

### LINKED ISSUE #5881: Save fuel state for next flight
### Aircraft Version

Experimental

### Description

Instead of the plane loading with a set amount of fuel on every load, it should save the fuel quantity from the previous flight and use that as the amount of fuel the plane loads in with. It would be nice to have as it feels weird having to unload fuel every time. 

Maybe it could be done similar to the way settings are saved in the EFB although I'm not too sure how it would be done or if it can be done. 

### References (optional)

_No response_

### Additional info (optional)

_No response_

### Discord Username (optional)

oim#0001

### PR DESCRIPTION
**Pull request by juliansebline**
Fixes #5881

## Summary of Changes
This PR makes the sim able to start a flight with the fuel of the previous one.

## Additional context

A engine configuration is saved in directory _AircraftStates_ a file named with _AircraftRegistration.ini_ where AircraftRegistration is the registration found in _atc_id_ in _acft.cfg_. The default value is _ASX320_ if not found. 

**The file is created/updated ONLY on ground and with at least an engine off.** 

**AircraftStates/ASX320.ini example**

[fuel]
fuel_center_quantity = 0.000000
fuel_left_quantity = 364.437298
fuel_right_quantity = 343.734621
fuel_left_aux_quantity = 228.000000
fuel_right_aux_quantity = 228.000000

## Contact
Julian Sebline (Julian Sebline#8476)
Thanks to _TazX [Z+2]#0405_ and _someperson#4953_ for the help!

## Testing instructions

**INITIAL REQUIREMENTS**

1. Close MSFS
2. Start Simvars Watcher
   1. Set FUEL TANK LEFT/RIGHT MAIN QUANTITY and FUEL TANK LEFT/RIGHT AUX QUANTITY and FUEL TANK CENTER QUANTITY
3. Open the /work directory
   1. The directory _AircraftStates_ must not exist 

**FIRST TEST: Save fuel into a file AND read a file and set the fuel**

1. Start a flight on the runway
   1. Default values should be main/aux/center : 400/228/0 
2. Set full thrust on one engine to increase the fuel consumption (to build up a fair difference) and wait tens of seconds
4. Turn off one engine
5. Check /work directory
    1.  The directory _AircraftStates_ with _A32NX.ini_ should exists
    2. Its content should be as in "Additional context" section
6. Restart the flight
   1. The quantities should be the same 

**SECOND TEST: Fuel not saved in flight**

1. Return to main menu
3. Start a flight in the air
4. Shutdown one engine
5. Check /work directory
   1. Directory _AircraftStates_ OR the file _A32NX.ini_ should not exist OR the values are never updated (because we don't want to save the fuel in flight)

**THIRD TEST: Change of _atc_id_ took into account**

1. Close MSFS
2. Change _atc_id_ in _acft.cfg_ (there are two files like this on my system) for a string of your choice
3. Do again the FIRST TEST
4. The configuration file should be named _StringOfYourChoice.ini_

**FOURTH TEST: Default registration when _atc_id_ not found within _acft.cfg_**

1. Close MSFS
2. Comment out _atc_id_ in _acft.cfg_ (there are two files like this on my system)
3. Do again the FIRST TEST
4. The configuration file should be named _ASX320.ini_ (as ASX320 is the default registration)

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

**Comment by Gurgel100** [comment]
I didn't reviewed your code. But I think it would be better to create a new directory in the work directory called "AircraftStates" or something like that where you store the state file as AircraftRegistration.ini. This way other system states can be stored in the same file later on and you will not have a lot of files inside the work directory. Also makes it easier to backup if you want to clear the work directory.

**Comment by juliansebline** [comment]
Therefore each section within the file would be a different livery ?

**Comment by Gurgel100** [comment]
No each livery would still have its own file. But different systems would have a different section inside the file.

**Comment by juliansebline** [comment]
Ah sorry, I had to replace AircraftRegistration with a real registration, my apologies.

Therefore it would be the same format of file but with a shorter name and in a different directory.

I can do that very quickly tomorrow

**Comment by Taz5150** [comment]
Going over the code changes in the FADEC, just a quick note to let everyone know there are no potential interference with the engine parameters or functionality. Thumbs up from my side.

File: src/fbw/src/interface/SimConnectInterface.cpp [hunk 1]
```diff
@@ -242,6 +242,8 @@ bool SimConnectInterface::prepareSimDataSimConnectDataDefinitions() {
   result &= addDataDefinition(hSimConnect, 0, SIMCONNECT_DATATYPE_FLOAT64, "FLAPS HANDLE INDEX", "NUMBER");
   result &= addDataDefinition(hSimConnect, 0, SIMCONNECT_DATATYPE_FLOAT64, "GEAR HANDLE POSITION", "POSITION");
 
+  result &= addDataDefinition(hSimConnect, 8, SIMCONNECT_DATATYPE_STRING32, "ATC ID", "");
```
Comment (aguther) [inline src/fbw/src/interface/SimConnectInterface.cpp:245]: This should not be added here as it is not being used.

File: src/fbw/src/interface/SimConnectInterface.cpp [hunk 2]
```diff
@@ -2533,6 +2535,7 @@ bool SimConnectInterface::isSimConnectDataTypeStruct(SIMCONNECT_DATATYPE type) {
     case SIMCONNECT_DATATYPE_WAYPOINT:
     case SIMCONNECT_DATATYPE_LATLONALT:
     case SIMCONNECT_DATATYPE_XYZ:
+    case SIMCONNECT_DATATYPE_STRING32:
```
Comment (aguther) [inline src/fbw/src/interface/SimConnectInterface.cpp:2538]: Can also be removed.

**Review by aguther — CHANGES_REQUESTED** [review]

File: src/fbw/src/interface/SimConnectInterface.cpp (hunk 2 above)
Comment (juliansebline) [reply src/fbw/src/interface/SimConnectInterface.cpp]: Fixed in a new commit

File: src/fbw/src/interface/SimConnectInterface.cpp (hunk 1 above)
Comment (juliansebline) [reply src/fbw/src/interface/SimConnectInterface.cpp]: Fixed in a new commit

**Comment by juliansebline** [comment]
Thanks to everyone for your help

**Review by aguther — APPROVED** [review]

**Comment by mico975** [comment]
Quality Assurance Tester Report

Discord          : mico#3145
Object of testing: #7170
Tier of Testing  : 1
Date             : 13/05/2022

Testing Process: 
FIRST TEST: Save fuel into a file AND read a file and set the fuel
Start a flight on the runway
Default values should be main/aux/center : 400/228/0 - **OK**
Set full thrust on one engine to increase the fuel consumption (to build up a fair difference) and wait tens of seconds - **OK**
Turn off one engine
Check /work directory
The directory AircraftStates with A32NX.ini should exists - **OK**
Its content should be as in "Additional context" section
Restart the flight
The quantities should be the same - **OK**

SECOND TEST: Fuel not saved in flight

Return to main menu
Start a flight in the air
Shutdown one engine
Check /work directory
Directory AircraftStates OR the file A32NX.ini should not exist OR the values are never updated - **OK**

THIRD TEST: Change of atc_id took into account
Close MSFS
Change atc_id in acft.cfg (there are two files like this on my system) for a string of your choice
Do again the FIRST TEST - **OK**
The configuration file should be named StringOfYourChoice.ini - **OK**

FOURTH TEST: Default registration when atc_id not found within acft.cfg
Close MSFS
Comment out atc_id in acft.cfg (there are two files like this on my system)
Do again the FIRST TEST - **OK**
The configuration file should be named ASX320.ini - **Not OK**, it somehow names it "Â " on my system

Negatives:       
Small negative with file naming, otherwise great.

Testing Results: 
Passed

Conclusions: 
LGTM

**Comment by Benjozork** [comment]
For future reference, this is a feature, not a fix.

**Pull request merged**